Add documentation and plugin notice about the end of the plugin version

Levdbas opened the pull request:
Related:

- #2794

## Work in progress
This is a first version of this PR where I would like to have input from the core team: @gchtr , @nlemoine and @jarednova . Could you have a look at the wording of the notices in the .txt, .md file? Is that what we want to communicate? Please change/complete my suggested texts and notifications where necessary so that we can release this.

**Example of the plugin notice as currently commited**
![afbeelding](https://github.com/timber/timber/assets/17764157/1f3b558d-0eb2-45b4-aace-1c18aae67fad)


### TODO's

- [x] Fill the links in the code changes
- [x] test the upgrade guide with a minimal theme and latest WordPress version
- [x] [Release a plugin version with the new code changes.](https://github.com/timber/timber/issues/2820)


## Issue
We are not going to release a plugin version of Timber anymore. We think that a lot of people will be stuck on the plugin version of Timber 1. With a install base of 50.000+ installations we are talking about a lot of people.


## Solution
Let's inform Timber users about the upcoming end of the Timber plugin and help them with extra guides on how to switch to the Composer based (1.x) version.


## Impact
Minimal on the codebase. This PR adds an inline plugin notification as the only code change.


## Usage Changes
See the whole guide :)


## Considerations

- I made a docs/getting-started/switch-to-composer.md guide. Is this the proper location?
- See the plugin notification I currently created. Is this the way we want to notify the plugin users about this change
- Is the switch-to-composer guide good enough?



## Testing
No

Levdbas committed:
docs: Add documentation about the Timber plugin

Levdbas committed:
(feat) Add deprecation notice for the plugin notice

jrathert commented on README.md:
```diff
@@ -19,6 +19,15 @@ By
 [![!Financial Contributors](https://opencollective.com/timber/tiers/badge.svg)](https://opencollective.com/timber/)
 
+> ⚠️ **Important information about the Timber plugin**
+> With the upcoming release of Timber 2, we will stop releasing Timber as a plugin. We advice everyone to switch to the composer based install as soon as possible.
```
I guess we should say something about the fact that v1 will no more supported with PHP 8.2 (has this been decided/announced?).

gchtr replied:
Yes, this is important to note and one of the main reasons why people should make the switch, too.

Levdbas replied:
I tried to add it to the notice as well. @jrathert @gchtr. I also clearified that the upgrade guide is only the first step in upgrading to Timber 2.0

jrathert commented on docs/getting-started/setup.md:
```diff
@@ -7,12 +7,20 @@ menu:
 
 ## Installation
+> ⚠️ **Important information**
+> With the upcoming release of Timber 2, we will stop releasing Timber as a plugin. We advice everyone to switch to the composer based install as soon as possible.
```
See above - say something about PHP 8.2

jrathert commented on docs/getting-started/switch-to-composer.md:
````diff
@@ -0,0 +1,65 @@
+Now that the plugin is disabled, you can install the latest 1.x version of Timber via composer. You can do this by navigating to your site's active theme folder inside your terminal and then running the following commands:
+
+```shell
+composer init
````
This init call and answering the questions is not needed at all. You just need to issue the "require" line to fetch the right files.

jrathert commented on docs/getting-started/switch-to-composer.md:
````diff
@@ -0,0 +1,65 @@
+This will start the composer initialization process. You can just press enter on all the questions that are asked. This will create a `composer.json` file in your theme folder. After that we can require the latest 1.x version of Timber by running the following command:
+
+```shell
+composer require timber/timber:^1.0
````
Maybe add a comment that composer might ask you to allow plugins and it is safe to say "yes"? Would help people with little composer know-how (like me ;-).

Levdbas replied:
Great suggestion. Added that as well!

jrathert commented on docs/getting-started/switch-to-composer.md:
````diff
@@ -0,0 +1,65 @@
+```
+
+### 5. Load the composer autoloader and initialize Timber
+Now that we have Timber installed via composer, we need to load the composer autoloader and initialize Timber. You can do this by adding the following code at the top of your `functions.php` file:
````
Maybe add that the following lines might actually _already be_ in your functions.php (i.e., if your theme is based on the Timber starter theme)

jrathert commented on docs/getting-started/switch-to-composer.md:
```diff
@@ -0,0 +1,65 @@
+Now that you have Timber installed via composer, you can check if your website still runs as expected. If you run into any issues, please try to solve them yourself by activating WordPress debugging and checking the error logs.
+
+### 7. Deploy your changes to your live website
+When you are 100% sure that you have succesfully upgraded your website to the composer based version of Timber, you can disable the Timber plugin, deploy your theme changes to your live website. If you are using a deployment pipeline, you can commit your changes and push them to your live website and make sure that the new `vendor` folder inside your theme gets deployed as well. If you are not us ...
```
May I suggest the following changes: 
When you are 100% sure that you have successfully upgraded your theme in the development website to the composer based version of Timber, you can deploy your theme changes to your live website. Make sure that the new `vendor` folder inside your theme gets deployed as well. If you are using a deployment pipeline, you can commit your changes and push them to your live website. If you are not using a version control system, you can use SCP or FTP to upload your changes to your theme. As a last step you should disable and then remove the Timber v1 plugin from your live website. Congrats - you successfully moved to the composer-based install and are prepared to later upgrade to Timber v2.

jrathert commented on docs/getting-started/setup.md:
```diff
@@ -7,12 +7,20 @@ menu:
+
+### ~~Via WordPress.org (easy)~~
 
 You can grab the all-things-included plugin at [WordPress.org](http://wordpress.org/plugins/timber-library/) either through the WordPress site or through the Plugins menu in the backend. Then skip ahead to [using the starter theme](#use-the-starter-theme).
```
Would take these lines out - we should simply not suggest and describe how to use the plugin...

gchtr replied:
```suggestion
~~You can grab the all-things-included plugin at [WordPress.org](http://wordpress.org/plugins/timber-library/) either through the WordPress site or through the Plugins menu in the backend. Then skip ahead to [using the starter theme](#use-the-starter-theme).~~
```

Levdbas commented:
Thank you for the feedback @jrathert, I'll pick this up very soon.

gchtr commented on readme.txt:
```diff
@@ -11,6 +11,14 @@ License URI: http://www.gnu.org/licenses/gpl-2.0.html
 
 == Description ==
+With the upcoming release of Timber 2, we will not release a 2.0 version and beyond  in plugin form. We advice everyone to switch to the composer based install as soon as possible.
+You will find an extensive list with guides and the reasons why we are not going to release Timber 2 in plugin version anymore.
```
```suggestion
You will find an extensive list with guides and the reasons why we are not going to release Timber 2.0 as a plugin anymore.
```

gchtr commented on readme.txt:
```diff
@@ -11,6 +11,14 @@ License URI: http://www.gnu.org/licenses/gpl-2.0.html
+With the upcoming release of Timber 2, we will not release a 2.0 version and beyond  in plugin form. We advice everyone to switch to the composer based install as soon as possible.
+You will find an extensive list with guides and the reasons why we are not going to release Timber 2 in plugin version anymore.
+
+### Switching to the composer based version
```
```suggestion
### Switching to the Composer based version
```

gchtr commented on readme.txt:
```diff
@@ -11,6 +11,14 @@ License URI: http://www.gnu.org/licenses/gpl-2.0.html
+You will find an extensive list with guides and the reasons why we are not going to release Timber 2 in plugin version anymore.
+
+### Switching to the composer based version
+* Guide: [How do I switch over from the plugin version to the composer based version of Timber?](https://timber.github.io/docs/getting-started/switch-to-composer/)
```
```suggestion
* Guide: [How do I switch over from the plugin version to the Composer based version of Timber?](https://timber.github.io/docs/getting-started/switch-to-composer/)
```

gchtr commented on readme.txt:
```diff
@@ -11,6 +11,14 @@ License URI: http://www.gnu.org/licenses/gpl-2.0.html
+### Switching to the composer based version
+* Guide: [How do I switch over from the plugin version to the composer based version of Timber?](https://timber.github.io/docs/getting-started/switch-to-composer/)
+* Backstory: [Why are we dropping support for the plugin in the first place](https://github.com/timber/timber/pull/2005)
+* Github issue: [Road to Timber 2.0](https://github.com/timber/timber/issues/2741)
```
```suggestion
* GitHub issue: [Roadmap for Timber 2.0](https://github.com/timber/timber/issues/2741)
```

gchtr commented on docs/getting-started/setup.md:
```diff
@@ -7,12 +7,20 @@ menu:
-### Via WordPress.org (easy)
+Underneath this text you will find an extensive list with guides and the reasons why we are not going to release Timber 2 in plugin version anymore.
+
+* Guide: [How do I switch over from the plugin version to the composer based version of Timber?](https://timber.github.io/docs/getting-started/switch-to-composer/)
```
Add small updates from above.

gchtr commented on lib/Admin.php:
```diff
@@ -6,13 +6,25 @@ class Admin {
 
 	public static function init() {
 		$filter = add_filter('plugin_row_meta', array(__CLASS__, 'meta_links'), 10, 2);
+		$action = add_action('after_plugin_row_timber-library/timber.php', array(__CLASS__, 'show_deprecation_warning'), 10, 3);
```
Yes, this is exactly what I meant. This catches the attention of people who are responsible for managing the plugins without being too intrusive.

gchtr commented on lib/Admin.php:
```diff
@@ -6,13 +6,25 @@ class Admin {
+	 */
+	public static function show_deprecation_warning($file, $plugin, $status)
+	{
+		echo '<td colspan="3" class="plugin-update colspanchange"><div class="update-message notice-error notice inline notice-warning notice-alt"><p>We recommend <a target="_blank" href="#">switching to the Composer based version of Timber</a>. <a href="#" target="_blank">Why you need to switch.</a></p></div></td>';
```
I’d make the first link take fewer words so that it’s easier to see that it’s two links.

```suggestion
		echo '<td colspan="3" class="plugin-update colspanchange"><div class="update-message notice-error notice inline notice-warning notice-alt"><p>We recommend <a target="_blank" href="#">switching to the Composer based version</a> of Timber. <a href="#" target="_blank">Why you need to switch.</a></p></div></td>';
```

Levdbas replied:
Nice, added the link as well @gchtr

gchtr reviewed (commented):
This is fantastic, @Levdbas. Thanks so much, I’m sure this will help so many people to make the switch.

> I made a docs/getting-started/switch-to-composer.md guide. Is this the proper location?

Yes, this looks good to me!

> See the plugin notification I currently created. Is this the way we want to notify the plugin users about this change

Yes, I would have done it like that.

> Is the switch-to-composer guide good enough?

Oh yes it is.

Thanks also to @jrathert for taking the time to read through this as someone who hasn’t a lot of experience with Composer 🙌.

Levdbas committed:
Applied suggestions of @gchtr

Co-authored-by: Lukas Gächter <lukas.gaechter@mind.ch>

Levdbas commented on README.md:
```diff
@@ -19,6 +19,15 @@ By
 [![!Financial Contributors](https://opencollective.com/timber/tiers/badge.svg)](https://opencollective.com/timber/)
 
+> ⚠️ **Important information about the Timber plugin**
+> With the upcoming release of Timber 2, we will stop releasing Timber as a plugin. We advice everyone to switch to the composer based install as soon as possible.
```
Are we calling it Timber 2 or Timber 2.0? Added a link as well.

Levdbas committed:
Updated readme.md

Levdbas committed:
Apply suggestions to setup guide

Levdbas committed:
updated admin notice

Levdbas committed:
Applied suggestions of @gchtr and @jrathert

Also clearified the intro paragraph that this guide is for switching to v1.

Levdbas committed:
added captials

Levdbas committed:
and another capital

Levdbas committed:
add additional link

Levdbas committed:
add announcements as well

Levdbas commented:
Hi @jrathert  and @gchtr, thanks for all the suggestions. I _think_ I added them all and fixed up a final few todos as well. Maybe lets have one more pass at this?

jrathert commented on docs/getting-started/switch-to-composer.md:
```diff
@@ -0,0 +1,76 @@
+
+4. Install the latest 1.x version of Timber via Composer
+5. Load the Composer autoloader and initialize Timber
+5. Load the composer autoloader and initialize Timber 
```
```suggestion
```

jrathert commented on docs/getting-started/switch-to-composer.md:
```diff
@@ -0,0 +1,76 @@
+
+1. Get a local development version of your website up and running.
+2. Check if you are running the latest version of the Timber plugin. If not, update it and check if your website still runs as expected.
+
```
```suggestion
3. Disable the Timber plugin
```

jrathert commented on docs/getting-started/switch-to-composer.md:
```diff
@@ -0,0 +1,76 @@
+6. Check if your website still runs as expected
+7. Deploy your changes to your live website
+
+We highly recommend doing these steps on a local development version of your website. If you don’t have one yet, you could use [Local by Flywheel](https://localbyflywheel.com/) to get one up and running in a few minutes.
```
```suggestion
### 1. Get a local development version of your website up and running
We highly recommend doing these steps on a local development version of your website. If you don’t have one yet, you could use [Local by Flywheel](https://localbyflywheel.com/) to get one up and running in a few minutes.
```

jrathert commented on docs/getting-started/switch-to-composer.md:
```diff
@@ -0,0 +1,76 @@
+
+### 3. Disable the Timber plugin
+Once you are sure that you are running the latest version of the Timber plugin, you can disable it. This will make sure that the plugin is not loaded anymore and does not interfere with the Composer based version of Timber.
+
```
```suggestion

Please note that your website will at this point in time no more work as there is no Timber available. This will be fixed in the next step.
```

Levdbas replied:
Good addition, I changed it a bit to:

`Please note that you website will throw erros at this point, as there is no Timber available. This will be fixed in the next two steps.`

jrathert commented on docs/getting-started/switch-to-composer.md:
````diff
@@ -0,0 +1,76 @@
+
+```php
+/**
+ * If you are installing Timber as a Composer dependency in your theme, you'll need this block
````
```suggestion
 * As Timber is installed as a Composer dependency, we need this block
```

jrathert commented on docs/getting-started/switch-to-composer.md:
````diff
@@ -0,0 +1,76 @@
+```php
+/**
+ * If you are installing Timber as a Composer dependency in your theme, you'll need this block
+ * to load your dependencies and initialize Timber.
````
```suggestion
 * to load dependencies and initialize Timber.
```

Levdbas replied:
I have redone this codeblock to be more in line with the function.php of the [Timber starter theme v2 functions file](https://github.com/timber/starter-theme/blob/2.x/functions.php)

jrathert reviewed (changes requested):
Hi @Levdbas - I added a few suggestions (some of the last changes were actually erroneous, I think). Thanks!

jrathert commented:
> Maybe lets have one more pass at this?

@Levdbas , thanks for the hard work (I am not only new to PHP, Twig, Timber and Composer, but also not too "github-savy" yet, but I keep on learning - I tried to add a few smaller changes/fixes as a review, hope it worked... ;-))

One thing that came to my mind when reading #2804: We always say we drop support for the plugin version of Timber (v1, needless to say). But will we still support the Composer-version of Timber v1? E.g., are there any plans/decision re: updating the code of v1 to make it PHP 8.2-friendlier at least (see #2793)? 

Because I fear if we only tell people now to switch to the Composer version, they might feel safe re: PHP 8.2 as well, and maybe miss that they _need_ to upgrade/change something one more time.

Levdbas committed:
(docs) Added suggestions from @jrathert

Levdbas committed:
Fixes wording as suggested by @gchtr

nlemoine reviewed (dismissed):
Great work @Levdbas !

jarednova commented:
This is looking good @Levdbas! Can you tag me when this is merged and ready for deployment to wp.org? Also happy to invite your wp.org user name if you're comfortable inside their SVN environment

Levdbas commented:
I think we are good to go here! Although I have my own plugin in the .org repo, to say that I am comfortable with the SVN stuff, not really... 👯 

The path forward would be:

1. Merge PR and maybe directly https://github.com/timber/timber/pull/2805 as well 
2. Deploy new version of the docs
3. Check if the links work as intended.
4. Release new plugin version, while at it, maybe update the "tested up to" version as well?
5. The first two steps of this issue: https://github.com/timber/timber/issues/2794 @jarednova

jarednova committed:
tweaks to readme style and link

jarednova reviewed (dismissed):
This looks great! I just adjusted the emphasis on some of the language at top (emphasizing Composer as the supported method) and tightening the formatting a bit

jarednova commented:
@Levdbas I feel like you have the best view into sequencing of how we merge this (#2800), #2805 and #2794. Could you proceed with merging this into master when you have the necessary input and dependencies?

Once merged, please create a 1.x release checklist ...

https://github.com/timber/timber/issues/new?assignees=&labels=&projects=&template=release_checklist_1x.md

That will be my cue that things are ready for SVN ~hell~ delight!

jarednova committed:
Update setup.md

Applies readme language to setup.md doc

jarednova committed:
Update switch-to-composer.md

Adapt language for Switch to Composer doc

gchtr commented:
> And regarding the support of the Composer Timber v1 version. I am pretty sure we are not going to actively update it anymore. So to make it compatible with PHP 8.2 would be out of scope as well.
> 
> Going though the documentation of this PR and already existing documentation, how would you suggest to communicate that as well? @gchtr , can you chime in on this as well?

I’d say at the moment, we won’t maintain the 1.x version anymore than really necessary.



> The path forward would be:
> 
> 1. Merge PR and maybe directly [docs: updated plugin support part #2805](https://github.com/timber/timber/pull/2805) as well
> 2. Deploy new version of the docs

I have an addition to point 2 in this list: We need to merge this into 2.x as well before we can deploy the docs. Because the docs are built against the 2.x branch.

@Levdbas I’ll let you merge this when you feel that this is ready. From my side, this looks all good. After merging this PR, I can see that we can merge it into 2.x as well and publish a new version of the docs.

Levdbas commented:
Alright @gchtr , Its merged! So you will merge the changes made to the MD files into the 2.x branch?

aevansgit commented:
I'm using Timber v1.23 on a site that we're planning to replace with a new build but not until June or July 2024. Do you think I'll be able to get away with leaving as is?

Levdbas commented:
Hi @aevansgit , are you using the plugin version? If so, you could try to migrate to the composer based version of 1.23. This way you will be able to receive one final update for the 1.x branch which will improve compatibility with PHP 8.1.

See https://timber.github.io/docs/v1/getting-started/switch-to-composer/
